### ISSUE TITLE
[BUG] Acer Swift Go 14 - No sound card detected - "No SoundWire machine driver found"

### ISSUE THREAD

**Issue by rodrigovivi** [thesofproject/linux#4923]
**Describe the bug**
No sound card detected on a laptop - Acer Swift Go 14

System : Fedora 40

**Packages** :
Fedora:
alsa-sof-firmware-2024.03-2.fc40.noarch
alsa-firmware-1.2.4-12.fc40.noarch

And even with manually installing: sof-bin-2024.03.tar.gz

**To Reproduce**
Install Linux on this Acer laptop.

sudo modprobe -r snd_sof_pci_intel_mtl
sudo modprobe snd_sof_pci_intel_mtl

**Reproduction Rate**
All the time

**Expected behavior**
The sound card should be detected.

**Impact**
showstopper
It is necessary for meeting.

**Environment**

Branch name and commit hash of the 2 repositories: sof (firmware/topology) and linux (kernel driver).
Kernel: : vmlinuz-6.8.5-300.fc40.x86_64 or Linus 6.9.0-rc3+
SOF:  sof-bin-2024.03
Name of the topology file
Topology file: intel/sof-ace-tplg/sof-hda-generic-idisp-2ch.tplg
Name of the platform(s) on which the bug is observed.
Platform: Acer Swift Go 14
Screenshots or console output

**  No SoundWire machine driver found **

```
[  119.281924] sof-audio-pci-intel-mtl 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if info 0x040100
[  119.281952] sof-audio-pci-intel-mtl 0000:00:1f.3: SoundWire enabled on CannonLake+ platform, using SOF driver
[  119.282189] sof-audio-pci-intel-mtl 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if 0x040100
[  119.282297] sof-audio-pci-intel-mtl 0000:00:1f.3: bound 0000:00:02.0 (ops i915_audio_component_bind_ops [i915])
[  119.289796] sof-audio-pci-intel-mtl 0000:00:1f.3: use msi interrupt mode
[  119.303215] sof-audio-pci-intel-mtl 0000:00:1f.3: No SoundWire machine driver found
[  119.303218] sof-audio-pci-intel-mtl 0000:00:1f.3: hda codecs found, mask 4
[  119.303220] sof-audio-pci-intel-mtl 0000:00:1f.3: using HDA machine driver skl_hda_dsp_generic now
[  119.303224] sof-audio-pci-intel-mtl 0000:00:1f.3: DMICs detected in NHLT tables: 2
[  119.304432] sof-audio-pci-intel-mtl 0000:00:1f.3: SOF firmware and/or topology file not found.
[  119.304434] sof-audio-pci-intel-mtl 0000:00:1f.3: Supported default profiles
[  119.304435] sof-audio-pci-intel-mtl 0000:00:1f.3: - ipc type 1 (Requested):
[  119.304437] sof-audio-pci-intel-mtl 0000:00:1f.3:  Firmware file: intel/sof-ipc4/mtl/sof-mtl.ri
[  119.304438] sof-audio-pci-intel-mtl 0000:00:1f.3:  Topology file: intel/sof-ace-tplg/sof-hda-generic-idisp-2ch.tplg
[  119.304440] sof-audio-pci-intel-mtl 0000:00:1f.3: Check if you have 'sof-firmware' package installed.
[  119.304441] sof-audio-pci-intel-mtl 0000:00:1f.3: Optionally it can be manually downloaded from:
[  119.304442] sof-audio-pci-intel-mtl 0000:00:1f.3:    https://github.com/thesofproject/sof-bin/
[  119.304470] sof-audio-pci-intel-mtl 0000:00:1f.3: Direct firmware load for (null)/(null) failed with error -2
[  119.304473] sof-audio-pci-intel-mtl 0000:00:1f.3: error: sof firmware file is missing, you might need to
[  119.304474] sof-audio-pci-intel-mtl 0000:00:1f.3:        download it from https://github.com/thesofproject/sof-bin/
[  119.304475] sof-audio-pci-intel-mtl 0000:00:1f.3: error: failed to load DSP firmware -2
[  119.306005] sof-audio-pci-intel-mtl 0000:00:1f.3: error: sof_probe_work failed err: -2
```

[alsa-info.txt](https://github.com/thesofproject/linux/files/14962507/alsa-info.txt)

**Comment by plbossart** [comment]
@rodrigovivi if you have 2 DMICs and no HDaudio external codec, then chances are the configuration is similar to the HP SoundWire versions.

Can you add this file 
[sof-dyndbg.conf.txt](https://github.com/thesofproject/linux/files/14962935/sof-dyndbg.conf.txt) as /etc/modprobe.d/sof-dyndbg.conf, reboot then attach the dmesg log.

Also provide the result of 
````
cat /sys/bus/acpi/devices/*/adr  | grep 025d
````

**Comment by rodrigovivi** [comment]
$ cat /sys/bus/acpi/devices/*/adr  | grep 025d
0x000030025d071201

**Comment by rodrigovivi** [comment]
[dmesg-audio-dyndbg-config.txt](https://github.com/thesofproject/linux/files/14963628/dmesg-audio-dyndbg-config.txt)

**Comment by plbossart** [comment]
Thanks @rodrigovivi, so you have one RT712 on SoundWire link0, 2 DMICs and HDMI support with iDISP.

````
[   14.813241] soundwire_intel:sdw_intel_probe_controller: acpi PRP00001:00: Creating 4 SDW Link devices
[   14.813350] soundwire_bus:sdw_extract_slave_id: soundwire sdw-master-0-0: SDW Slave Addr: 30025d071201
[   14.813352] soundwire_bus:sdw_extract_slave_id: soundwire sdw-master-0-0: SDW Slave class_id 0x01, mfg_id 0x025d, part_id 0x0712, unique_id 0x0, version 0x3
[   14.813487] snd_sof_intel_hda:hda_codec_probe: sof-audio-pci-intel-mtl 0000:00:1f.3: HDA codec thesofproject/sof#2 probed OK: response: 8086281d
````

of course we don't support this configuration...

Usually we have a RT712 for jack/amp and the RT1712 for the DMIC. Here the latter seems replaced by the PCH-attached DMIC.

wow. I'll start cooking up a patch shortly.

**Comment by plbossart** [comment]
@rodrigovivi can you try the patch in PR thesofproject/sof#4925 and also unzip/copy this file 
[sof-mtl-rt712-l0-2ch.tplg.gz](https://github.com/thesofproject/linux/files/14963787/sof-mtl-rt712-l0-2ch.tplg.gz) in /lib/firmware/intel/sof-ipc4-tplg/

Reboot and attach the results. You should see a sound card if I didn't screw up the indices.

Thanks!

**Comment by rodrigovivi** [comment]
[dmesg-audio-with-patch-take-1.txt](https://github.com/thesofproject/linux/files/14964891/dmesg-audio-with-patch-take-1.txt)
It goes further, then it fails...
[   16.204785] snd_sof:sof_route_load: sof-audio-pci-intel-mtl 0000:00:1f.3: sink module-copier.101.2 control none source eqiir.101.1
[   16.204786] sof_sdw sof_sdw: ASoC: physical link dmic01 (id 4) not exist
[   16.204787] sof_sdw sof_sdw: ASoC: topology: could not load header: -22
[   16.204832] sof-audio-pci-intel-mtl 0000:00:1f.3: error: tplg component load failed -22
[   16.204834] sof-audio-pci-intel-mtl 0000:00:1f.3: error: failed to load DSP topology -22
[   16.204835] sof-audio-pci-intel-mtl 0000:00:1f.3: ASoC: error at snd_soc_component_probe on 0000:00:1f.3: -22
[   16.204845] sof_sdw sof_sdw: ASoC: failed to instantiate card -22
[   16.204882] sof_sdw sof_sdw: error -EINVAL: snd_soc_register_card failed -22
[   16.204888] sof_sdw sof_sdw: probe with driver sof_sdw failed with error -22

**Comment by plbossart** [comment]
Ok that's probably the topology file, I don't really understand how the indices are handled. @bardliao can you take a look?

**Comment by rodrigovivi** [comment]
one important information here: the sound over USB works pretty well even without this patch.
USB speaker and USB headset works well. only the native laptop speakers and mic doesn't work.
I was surprised by that over the weekend. So, I really can do meetings on this laptop and sound works for my regular case.
But I'm still up to test any patch here and get this product really working out of the box. There might be users out there who really depend on the native speakers/mic.

**Comment by bardliao** [comment]
@rodrigovivi Can you try [sof-mtl-rt712-l0-2ch.zip](https://github.com/thesofproject/linux/files/14973475/sof-mtl-rt712-l0-2ch.zip)?

```
[   16.204349] snd_soc_sof_sdw:sof_card_dai_links_create: sof_sdw sof_sdw: sdw 3, ssp 0, dmic 2, hdmi 3, bt: 0
[   16.204350] snd_soc_sof_sdw:init_dai_link: sof_sdw sof_sdw: create dai link SDW0-Playback-SimpleJack, id 0
[   16.204371] snd_soc_sof_sdw:init_dai_link: sof_sdw sof_sdw: create dai link SDW0-Capture-SimpleJack, id 1
[   16.204372] snd_soc_sof_sdw:init_dai_link: sof_sdw sof_sdw: create dai link SDW0-Playback-SmartAmp, id 2
[   16.204372] snd_soc_sof_sdw:init_dai_link: sof_sdw sof_sdw: create dai link dmic01, id 3
[   16.204373] snd_soc_sof_sdw:init_dai_link: sof_sdw sof_sdw: create dai link dmic16k, id 4
[   16.204375] snd_soc_sof_sdw:init_dai_link: sof_sdw sof_sdw: create dai link iDisp1, id 5
[   16.204376] snd_soc_sof_sdw:init_dai_link: sof_sdw sof_sdw: create dai link iDisp2, id 6
[   16.204377] snd_soc_sof_sdw:init_dai_link: sof_sdw sof_sdw: create dai link iDisp3, id 7
```

The dmic01 id is 3. I modified the indices in the topology.

**Comment by plbossart** [comment]
> The dmic01 id is 3. I modified the indices in the topology.

@bardliao is this because we don't have the amp feedback on rt712?

**Comment by plbossart** [comment]
> one important information here: the sound over USB works pretty well even without this patch. USB speaker and USB headset works well. only the native laptop speakers and mic doesn't work. I was surprised by that over the weekend. So, I really can do meetings on this laptop and sound works for my regular case. But I'm still up to test any patch here and get this product really working out of the box. There might be users out there who really depend on the native speakers/mic.

Yes, BT and USB are orthogonal to the Intel IP for audio, so they might work even if the DSP doesn't. Still there's no reason why this configuration wouldn't work with minor tweaks, as you said there will be lots of others who need jack/speaker/mic.

**Comment by bardliao** [comment]
> > The dmic01 id is 3. I modified the indices in the topology.
> 
> @bardliao is this because we don't have the amp feedback on rt712?

Yes.

**Comment by rodrigovivi** [comment]
with dmic01 id 3 it goes further, and then it fails:
[   18.577436] sof_sdw sof_sdw: ASoC: physical link dmic16k (id 5) not exist

[dmesg-audio-dmic01-id-3.txt](https://github.com/thesofproject/linux/files/14986066/dmesg-audio-dmic01-id-3.txt)

**Comment by bardliao** [comment]
> with dmic01 id 3 it goes further, and then it fails: [ 18.577436] sof_sdw sof_sdw: ASoC: physical link dmic16k (id 5) not exist
> 
> [dmesg-audio-dmic01-id-3.txt](https://github.com/thesofproject/linux/files/14986066/dmesg-audio-dmic01-id-3.txt)

@rodrigovivi Thanks for testing. Could you try [sof-mtl-rt712-l0-2ch.zip](https://github.com/thesofproject/linux/files/14987476/sof-mtl-rt712-l0-2ch.zip)?

**Comment by bardliao** [comment]
@rodrigovivi Do you have a chance to test [sof-mtl-rt712-l0-2ch.zip](https://github.com/thesofproject/linux/files/14987476/sof-mtl-rt712-l0-2ch.zip)? We will merge https://github.com/thesofproject/sof/pull/9041 if the test result is positive.

**Comment by franc5** [comment]
> @rodrigovivi Do you have a chance to test [sof-mtl-rt712-l0-2ch.zip](https://github.com/thesofproject/linux/files/14987476/sof-mtl-rt712-l0-2ch.zip)? We will merge [thesofproject/sof#9041](https://github.com/thesofproject/sof/pull/9041) if the test result is positive.

Hello everyone! I'm also having this issue in a Swift SFG14-72 running latest Ubuntu 24.04. I tried the linked topology and it didn't work for me. Is this ticket going to be continued in [issue 9085](https://github.com/thesofproject/linux/issues/4970)? Is there something I can help to debug this?

**Comment by bardliao** [comment]
> > @rodrigovivi Do you have a chance to test [sof-mtl-rt712-l0-2ch.zip](https://github.com/thesofproject/linux/files/14987476/sof-mtl-rt712-l0-2ch.zip)? We will merge [thesofproject/sof#9041](https://github.com/thesofproject/sof/pull/9041) if the test result is positive.
> 
> Hello everyone! I'm also having this issue in a Swift SFG14-72 running latest Ubuntu 24.04. I tried the linked topology and it didn't work for me. Is this ticket going to be continued in [issue 9085](https://github.com/thesofproject/linux/issues/4970)? Is there something I can help to debug this?

@franc5 You need to apply https://github.com/thesofproject/linux/pull/4925 to your kernel. And from https://github.com/thesofproject/linux/issues/4970, https://github.com/thesofproject/sof/pull/9041 should be good. The DSP panic looks like a different issue.

**Comment by rodrigovivi** [comment]
I'm sorry for the late reply here. With the latest [sof-mtl-rt712-l0-2ch.zip](https://github.com/thesofproject/linux/files/14987476/sof-mtl-rt712-l0-2ch.zip) I do can see the output audio enumerated, but no sound is audible when testing the speakers.
also no microfone enumerated.
[dmesg-apr-29.txt](https://github.com/thesofproject/linux/files/15150469/dmesg-apr-29.txt)

**Comment by plbossart** [comment]
@rodrigovivi I don't know what's going on with GitHub but your attached dmesg-apr-29 didn't go through.

**Comment by rodrigovivi** [comment]
[dmesg-apr-29.txt](https://github.com/thesofproject/linux/files/15150469/dmesg-apr-29.txt)

**Comment by plbossart** [comment]
@rodrigovivi I don't see anything wrong in the dmesg log.
The speakers probably miss a mixer config.
For the dmics, I don't really understand what you meant by "no microphone enumerated". what is the output of 'arecord -l' ?

**Comment by rodrigovivi** [comment]
The Gnome Settings shows the output device, but even with tuning the volume, nothing is audible.
But for the input devices it just shows: "No Input device"

$ arecord -l
**** List of CAPTURE Hardware Devices ****
card 1: sofsoundwire [sof-soundwire], device 1: Jack In (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: sofsoundwire [sof-soundwire], device 10: DMIC Raw (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0


$ amixer
Simple mixer control 'Master',0
  Capabilities: pvolume pswitch pswitch-joined
  Playback channels: Front Left - Front Right
  Limits: Playback 0 - 65536
  Mono:
  Front Left: Playback 49806 [76%] [on]
  Front Right: Playback 49806 [76%] [on]
Simple mixer control 'Capture',0
  Capabilities: cvolume cswitch cswitch-joined
  Capture channels: Front Left - Front Right
  Limits: Capture 0 - 65536
  Front Left: Capture 49806 [76%] [on]
  Front Right: Capture 49806 [76%] [on]

**Comment by plbossart** [comment]
that's because you don't have a UCM file.

Try capturing something with "arecord -Dhw:0,10 -c2 -r48000 -fS32_LE -d10 test.wav" and see if there's any sound coming in.

**Comment by rodrigovivi** [comment]
I couldn't hear anything... 
then I plugged the usb-audio, where I can hear stuff and I played it again and still couldn't hear anything.
so, still output is listed, but it doesn't work. input is not listed and even arecord directly didn't worked.

**Comment by plbossart** [comment]
"arecord directly didn't work"
Did you mean that the recorded file is silent, or that some sort of error happened.
It could be that we are using the 'wrong' set of channels, the interface supports 4 channels so if the OEM picks 2 we have a chance on using the pdm1 instead of classic pdm0.

**Comment by rodrigovivi** [comment]
silent! I couldn't listen to anything...

**Comment by plbossart** [comment]
ok, so we need to try the 4ch setup and see if the channels 2-3 are used 

Can you add this in a .conf file in /etc/modprobe.d/
````
options snd_sof_intel_hda_common dmic_num=4
````

and unzip/add this topology file 
[sof-mtl-rt712-l0-4ch.tplg.gz](https://github.com/thesofproject/linux/files/15152432/sof-mtl-rt712-l0-4ch.tplg.gz)

if you can record something and share the wav file that would be great.

**Comment by plbossart** [comment]
Oh and the number of channels needs to change of course:

arecord -Dhw:0,10 -c**4** -r48000 -fS32_LE -d10 test.wav

**Comment by rodrigovivi** [comment]
$ find /lib/firmware/intel/sof-ipc4-tplg/ | grep sof-mtl-rt712-l0-4ch
/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-4ch.tplg

$  cat /proc/cmdline 
BOOT_IMAGE=(hd0,gpt3)/vmlinuz-6.9.0-rc3+ root=/dev/mapper/fedora-root ro rd.lvm.lv=fedora/root rd.luks.uuid=luks-7ba20335-722d-4a79-9c82-451033b7ceb4 snd_sof_intel_hda_common.dmic_num=4 rhgb quiet

$ arecord -Dhw:0,10 -c4 -r48000 -fS32_LE -d10 test-4ch.wav
arecord: set_params:1398: Channels count non available

[dmesg-apr-30.txt](https://github.com/thesofproject/linux/files/15172710/dmesg-apr-30.txt)

**Comment by plbossart** [comment]
Not following what goes sideways, we do add the 4ch count
````
"cavs-sdw\;sof-mtl-rt712-l0-4ch\;PLATFORM=mtl,NUM_SDW_AMP_LINKS=1,\
SDW_AMP_FEEDBACK=false,SDW_SPK_STREAM=Playback-SmartAmp,\
SDW_JACK_OUT_STREAM=Playback-SimpleJack,SDW_JACK_IN_STREAM=Capture-SimpleJack,\
NUM_DMICS=4,PDM1_MIC_A_ENABLE=1,PDM1_MIC_B_ENABLE=1,DMIC0_ID=3,DMIC1_ID=4,\
PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-sof-mtl-rt712-l0-4ch.bin,\
HDMI1_ID=5,HDMI2_ID=6,HDMI3_ID=7"
````

It's possible though that somehow we use the NHLT blob coming from ACPI tables.

But at the same time, I see 
````
[  483.498873] snd_sof:sof_ipc4_prepare_copier_module: sof-audio-pci-intel-mtl 0000:00:1f.3: copier host-copier.10.capture: reference output rate 48000, channels 2
````
so not sure what's going on. @bardliao @ranj063 any ideas?

**Comment by bardliao** [comment]
Maybe try "options snd_sof_intel_hda_common sof_use_tplg_nhlt=1" to use the nhlt in the topology first?

**Comment by rodrigovivi** [comment]
options snd_sof_intel_hda_common sof_use_tplg_nhlt=1
and
options snd_sof_intel_hda_common dmic_num=4
?

**Comment by plbossart** [comment]
yes both options @rodrigovivi

**Comment by rodrigovivi** [comment]
Using both: $ cat /proc/cmdline 
[snip]
snd_sof_intel_hda_common.sof_use_tplg_nhlt=1 snd_sof_intel_hda_common.dmic_num=4

[dmesg-may13.txt](https://github.com/thesofproject/linux/files/15299548/dmesg-may13.txt)

$ arecord -Dhw:0,10 -c4 -r48000 -fS32_LE -d10 test-4ch-may-13.wav
arecord: set_params:1398: Channels count non available

mic still not listed in the gnome sound configuration.

**Comment by plbossart** [comment]
for some reason, even with the 4ch option the kernel selects the 2ch topology

````
[   15.253637] sof-audio-pci-intel-mtl 0000:00:1f.3: DMICs detected in NHLT tables: 2
[   15.253639] snd_sof_intel_hda_generic:hda_sdw_machine_select: sof-audio-pci-intel-mtl 0000:00:1f.3: SoundWire machine driver sof_sdw topology sof-mtl-rt712-l0-2ch.tplg
````

Not sure why that is the case.

My money is on the option not being selected, if you enable dynamic debug you should see this, but it's not in your log...

````c
/* allow for module parameter override */
	if (dmic_num_override != -1) {
		dev_dbg(sdev->dev,
			"overriding DMICs detected in NHLT tables %d by kernel param %d\n",
			dmic_num, dmic_num_override);
		dmic_num = dmic_num_override;
	}
````

**Comment by rodrigovivi** [comment]
what debug config and/or levels do you need?
But what to do to allow the override and not end up in this place.

**Comment by plbossart** [comment]
just make sure the sof-dyndbg.conf file exists in /etc/modprobe.d/ as described earlier in https://github.com/thesofproject/linux/issues/4923#issuecomment-2052436089

**Comment by marplusz** [comment]
I have the latest vanilla kernel for fedora:
uname -a
Linux fedora 6.11.0-0.rc0.20240724gt786c8248.311.vanilla.fc40.x86_64

I installed sof-bin v2024.06:

sudo find / | grep sof-mtl-rt712-l0
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-rt1712-l3.tplg.xz
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-rt1712-l3.tplg
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0.tplg

my laptop needs sof-mtl-rt712-l0-2ch.tplg:
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: enabling device (0000 -> 0002)
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if 0x040100
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: bound 0000:00:02.0 (ops i915_audio_component_bind_ops [i915])
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: use msi interrupt mode
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: DMICs detected in NHLT tables: 2
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: hda codecs found, mask 4
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: SOF firmware and/or topology file not found.
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: Supported default profiles
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: - ipc type 1 (Requested):
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3:  Firmware file: intel/sof-ipc4/mtl/sof-mtl.ri
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3:  Topology file: intel/sof-ace-tplg/sof-mtl-rt712-l0-2ch.tplg
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: Check if you have 'sof-firmware' package installed.
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: Optionally it can be manually downloaded from:
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3:    https://github.com/thesofproject/sof-bin/
lip 24 16:28:20 fedora kernel: sof-audio-pci-intel-mtl 0000:00:1f.3: error: sof_probe_work failed err: -2
lip 24 16:28:21 fedora bluetoothd[1109]: profiles/audio/micp.c:micp_init() D-Bus experimental not enabled

Where can I find sof-mtl-rt712-l0-2ch.tplg?

**Comment by plbossart** [comment]
@marplusz see https://github.com/thesofproject/linux/issues/4923#issuecomment-2054591643

I am not sure what the status is, we never had any test results on the files suggested in https://github.com/thesofproject/sof/pull/9041 and there are comments from @ujfalusi that were not addressed either.

Bottom line we need to reopen this for the 2ch case.

**Comment by marplusz** [comment]
I went through the entire thread and I have the same thing as @rodrigovivi

In the https://github.com/thesofproject/linux/files/14987476/sof-mtl-rt712-l0-2ch.zip version there was sound but only in the headphones jack.
No microphone.

sudo find / | grep sof-mtl-rt712-l0
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-rt1712-l3.tplg.xz
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-rt1712-l3.tplg
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0.tplg
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-2ch.tplg
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-4ch.tplg

arecord -Dhw:0,10 -c4 -r48000 -fS32_LE -d10 test.wav 
arecord: set_params:1398: Channels count non available

sudo find / | grep sof-dyndbg.conf
/etc/modprobe.d/sof-dyndbg.conf

Let me know how I can help you with your tests.

**Comment by bardliao** [comment]
@marplusz @rodrigovivi @plbossart The dmic_num=4 belongs to the snd_sof_intel_hda_generic module not snd_sof_intel_hda_common. So, the options should be

options snd_sof_intel_hda_common sof_use_tplg_nhlt=1
options **snd_sof_intel_hda_generic** dmic_num=4

And the tplg file will be `sof-mtl-rt712-l0-4ch.tplg`

**Comment by marplusz** [comment]
I set the options:
options snd_sof_intel_hda_common sof_use_tplg_nhlt=1
options snd_sof_intel_hda_generic dmic_num=4

I recorded a wav file:
arecord -Dhw:0,10 -c4 -r48000 -fS32_LE -d10 test.wav

No sound was recorded. I checked the file in Audacity and nothing was recorded on any channel.

**Comment by plbossart** [comment]
@marplusz can you add this file 
[sof-dyndbg.conf.txt](https://github.com/user-attachments/files/16374275/sof-dyndbg.conf.txt) as /etc/modprobe.d/sof-dyndbg.conf, reboot and attach the dmesg log. Thanks!

**Comment by marplusz** [comment]
> @marplusz can you add this file [sof-dyndbg.conf.txt](https://github.com/user-attachments/files/16374275/sof-dyndbg.conf.txt) as /etc/modprobe.d/sof-dyndbg.conf, reboot and attach the dmesg log. Thanks!

Log file from dmesg
[dmesg_log.txt](https://github.com/user-attachments/files/16377060/dmesg_log.txt)

**Comment by plbossart** [comment]
the log is missing all the initial parts that can help us see what the configuration is @marplusz

**Comment by marplusz** [comment]
@plbossart Please write what I should insert. Above I made the usual ones 
dmesg > dmesg_log.txt

**Comment by marplusz** [comment]
I updated the system and did it again. I can see the difference in the log. 
 I have a vanilla kernel so it may be unstable.
[dmesg_log2.txt](https://github.com/user-attachments/files/16382319/dmesg_log2.txt)

**Comment by bardliao** [comment]
@marplusz Can you share the output of "amixer contents" command while you are running the arecord command?

**Comment by marplusz** [comment]
@bardliao To do this, I restored the parameters: 
options snd_sof_intel_hda_common sof_use_tplg_nhlt=1 
options snd_sof_intel_hda_generic dmic_num=4 
I rebooted 
journalctl | grep audio -> journalctl.txt 
dmesg  -> dmesg_log

_terminal_1_ -> arecord -Dhw:0,10 -c4 -r48000 -fS32_LE -d10 test.wav 

_terminal_2_ -> amixer contents
numid=4,iface=MIXER,name='Master Playback Switch'
  ; type=BOOLEAN,access=rw------,values=1
  : values=on
numid=3,iface=MIXER,name='Master Playback Volume'
  ; type=INTEGER,access=rw------,values=2,min=0,max=655

36,step=1
  : values=45212,45212
numid=2,iface=MIXER,name='Capture Switch'
  ; type=BOOLEAN,access=rw------,values=1
  : values=on
numid=1,iface=MIXER,name='Capture Volume'
  ; type=INTEGER,access=rw------,values=2,min=0,max=65536,step=1
  : values=45212,45212
[journalctl.txt](https://github.com/user-attachments/files/16388183/journalctl.txt)
[dmesg_log.txt](https://github.com/user-attachments/files/16388237/dmesg_log.txt)

**Comment by bardliao** [comment]
@marplusz Sorry, you need to specify the card number. Please redo this with `amixer -c0 contents`.

**Comment by marplusz** [comment]
> @marplusz Sorry, you need to specify the card number. Please redo this with `amixer -c0 contents`.
@bardliao The file contains the result of the amixer -c0 contents command
[amixer.txt](https://github.com/user-attachments/files/16392306/amixer.txt)

**Comment by plbossart** [comment]
````numid=42,iface=MIXER,name='Dmic0 Capture Switch'
  ; type=BOOLEAN,access=rw------,values=4
  : values=off,off,off,off
````

That doesn't look so good. Try
````
amixer -Dhw:0 cset numid=42 on
````

**Comment by marplusz** [comment]
I did 
`amixer -Dhw:0 cset numid=42 on`
I have four channels in the wav file, of which the first two channels have sound from the microphone.

**Comment by plbossart** [comment]
ok, that looks encouraging @marplusz, can you try with this topology file
https://github.com/thesofproject/linux/files/14987476/sof-mtl-rt712-l0-2ch.zip

and change the options to 

````
options snd_sof_intel_hda_common sof_use_tplg_nhlt=1
options snd_sof_intel_hda_generic dmic_num=2
`````

and the command
````
arecord -Dhw:0,10 -c2 -r48000 -fS32_LE -d10 test.wav
````

Check if the mixers are set, not sure if they are persistent over reboot.

**Comment by marplusz** [comment]
These settings worked after reboot. 
The wav file has two channels and sound is recorded on both of them. 
I still have sound only on my headphones and Gnome settings don't see the microphone.

**Comment by plbossart** [comment]
@marplusz Gnome/PipeWire rely on UCM to show which devices are present. Make sure you have the latest of alsa-ucm-conf and please share the output of 'amixer -Dhw:0 info'. There should be something in the component string that mentions dmic numbers.

**Comment by marplusz** [comment]
I have Fedora 40 with the alsa-ucm-1.2.12-1.fc40.noarch package.  
amixer -Dhw:0 info
Card hw:0 'sofsoundwire'/'MTL-SwiftSFG14_72-V1.04-Coral_MTH'
  Mixer name	: 'Intel Meteor Lake HDMI'
  Components	: 'HDA:8086281d,80860101,00100000  cfg-amp:1 mic:dmic cfg-mics:2 iec61937-pcm:7,6,5 hs:rt712 spk:rt712'
  Controls      : 43
  Simple ctrls  : 19

**Comment by plbossart** [comment]
the component string shows the presence of 2 mics "cfg-mics:2"

In theory the UCM setup should detect this. Maybe try to use the upstream version of UCM, it could be that your version is a bit outdated (I don't track alsa-lib releases).

**Comment by marplusz** [comment]
I updated Alsa to the latest version and no changes. The microphone is still not visible. For me, it is important that it works, for example, in zoom. What about the sound from the speakers? I will not be able to test the sound via HDMI until two weeks later.

**Comment by plbossart** [comment]
@shumingfan @bardliao is there any UCM support for RT712? I see rt713-sdca.conf 

sof-soundwire.conf:             Regex "(cs42l43|rt5682|rt700|rt711|rt713(-sdca)?)"
sof-soundwire.conf:             Regex "(rt713-dmic|rt715(-sdca)?)"

so that won't work with the component string, and then all the mixers in those UCM files rely on the 'rt713' prefix but in this case the 'rt712' prefix is used, so it's a double "cannot possibly work"...

**Comment by shumingfan** [comment]
@rodrigovivi  Would you please test the UCM config below and share the test results? Thanks.
https://github.com/shumingfan/alsa-ucm-conf/commit/1c41851377090c1d252d8d7030fb82d7b603d3c9

**Comment by marplusz** [comment]
@shumingfan  I copied all the new and modified files and after restarting I don't even have sound on the headphones :(
I checked the configuration and I have a state consistent with the changes in this thread.

**Comment by shumingfan** [comment]
@rodrigovivi How about the speaker?
Could you input 'alsaucm reload' command to check whether the error log happens?
Do you see the endpoint device in the Sound settings?
And please share the 'alsa-info' file.

**Comment by marplusz** [comment]
@shumingfan 
1) alsaucm reload
sudo dmesg | grep alsa
There are no results. If you check somewhere else let me know.
2) No. Before the change I could see the speakers (although they didn't work) and the headphones when I plugged them in. Now I don't see any device. 
3) /usr/sbin/alsa-info.sh
[alsa-info.txt](https://github.com/user-attachments/files/16454080/alsa-info.txt)

**Comment by shumingfan** [comment]
@rodrigovivi 
I tried to make my environment close to yours.
The component string is below on my machine.
```
$ amixer -c0 info
Card sysdefault:0 'sofsoundwire'/'IntelCorporation-MeteorLakeClientPlatform-0.1-MTL_PDDR5SODIMMSBSRVP'
  Mixer name    : 'Intel Meteor Lake HDMI'
  Components    : 'HDA:8086281d,80860101,00100000  cfg-amp:1 mic:dmic cfg-mics:2 hs:rt712 spk:rt712'
  Controls      : 22
  Simple ctrls  : 16
```
The devices display correctly including the digital microphone.
Could you tar your ucm2 directory and share?

**Comment by marplusz** [comment]
> @rodrigovivi I tried to make my environment close to yours. The component string is below on my machine.
> 
> ```
> $ amixer -c0 info
> Card sysdefault:0 'sofsoundwire'/'IntelCorporation-MeteorLakeClientPlatform-0.1-MTL_PDDR5SODIMMSBSRVP'
>   Mixer name    : 'Intel Meteor Lake HDMI'
>   Components    : 'HDA:8086281d,80860101,00100000  cfg-amp:1 mic:dmic cfg-mics:2 hs:rt712 spk:rt712'
>   Controls      : 22
>   Simple ctrls  : 16
> ```


> 
> The devices display correctly including the digital microphone. Could you tar your ucm2 directory and share?

I think you wanted to tag me :)
[ucm2.tar.gz](https://github.com/user-attachments/files/16458366/ucm2.tar.gz)

**Comment by shumingfan** [comment]
@marplusz  I compared your ucm2 directory with my ucm2 directory. It is the same.
Could you reboot the system and dump the dmesg? Please share the kernel log.

**Comment by bardliao** [comment]
@marplusz Can you try the topology with 24 bit format?
[sof-mtl-rt712-l0-2ch-4ch.zip](https://github.com/user-attachments/files/16464721/sof-mtl-rt712-l0-2ch-4ch.zip)
There is a known issue when UCM try to open all PCMs in 32 bit format.

**Comment by marplusz** [comment]
@bardliao @shumingfan New topology files solved the problem :)
There is sound from the speaker, headphones and the microphone works.
[dmesg_log2.txt](https://github.com/user-attachments/files/16466817/dmesg_log2.txt)

**Comment by plbossart** [comment]
What issue were you referring to @bardliao ? In theory the problem only exists when we have 3+ streams in parallel with 32 bits each. I am not sure how this configuration would be reached by userspace?

**Comment by bardliao** [comment]
> What issue were you referring to @bardliao ? In theory the problem only exists when we have 3+ streams in parallel with 32 bits each. I am not sure how this configuration would be reached by userspace?

@plbossart https://github.com/thesofproject/sof/pull/9282 was merged recently. So, I guess the topology that you provided [sof-mtl-rt712-l0-2ch-4ch.zip](https://github.com/user-attachments/files/16464721/sof-mtl-rt712-l0-2ch-4ch.zip) is built without the PR which means 32 bit format will be used. Audio services will try to open all the listed PCM and it will result in no bandwidth error. I just rebuild the topology to make sure the topology uses 24 bit format.

**Comment by plbossart** [comment]
Ah yes, if there's Headphone playback, Headset Capture and Speaker playback all open in 32 bits then this would fail indeed.

**Comment by brentlu** [comment]
Just record that following configuration works on my side for HP/SPK/DP playback and HP/PCH DMIC recording.

OS: Ubuntu 24.04 LTS
linux: https://git.launchpad.net/~canonical-kernel/ubuntu/+source/linux-oem/+git/noble (tag: Ubuntu-oem-6.10-6.10.0-1006.6)
alsa-ucm-conf: https://github.com/shumingfan/alsa-ucm-conf (branch: add-rt712va)
alsa-lib: https://github.com/alsa-project/alsa-lib (commit: d9694398)
topology: sof-mtl-rt712-l0-2ch.tplg (found in above commit)

**Comment by marplusz** [comment]
From version 6.10 the kernel is ready, but there must be corrections in: 
**alsa-ucm-conf** 
**alsa-lib**
**sof** -> _sof-mtl-rt712-l0-2ch-4ch_
When can they be expected to be released so as not to perform these steps manually?

**Comment by falkanson** [comment]
Hello Guys, i appreciate the hard work you guys have all put into this...
Is there a way to install this early patch?  already tried installing the 6.11-rc but it seems like the fix has not yet made it into there...

I tried to match brentlu's confiq but i was pretty unsure on how to install the topology stuff, as that escaped my technical ability. I would appreciate the help!

**Comment by plbossart** [comment]
the kernel patch is in 6.10 already

````
Author: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>  2024-04-26 17:21:15
Committer: Mark Brown <broonie@kernel.org>  2024-04-29 17:10:00
Parent: 02e6f7cb487f18e1171ae6d12ad1066fbd25176d (ASoC: Intel: soc-acpi: mtl: add Dell SKU 0C64 and 0CC6)
Child:  6d339113df3ab510ce075a18ccb10a20cb325d4e (ASoC: Intel: soc-acpi-intel-lnl-match: adds RT714 and RT1318 support)
Branches: master, remotes/stable/linux-6.10.y and many more (168)
Follows: v6.9-rc2
Precedes: asoc-v6.10, sof-dev-rebase-20240503, sof-dev-rebase-20240509

    ASoC: Intel: soc-acpi: mtl: add support for Acer Swift Go 14
    
    This device has an RT712 on link0, but does not rely on RT1712 for the
    DMIC. PCH-attached DMICs are used instead.
    
    Closes: https://github.com/thesofproject/linux/issues/4923
    Reviewed-by: Bard Liao <yung-chuan.liao@linux.intel.com>
    Reviewed-by: Péter Ujfalusi <peter.ujfalusi@linux.intel.com>
    Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
    Link: https://lore.kernel.org/r/20240426152123.36284-5-pierre-louis.bossart@linux.intel.com
    Signed-off-by: Mark Brown <broonie@kernel.org>
````

the topology file can be installed in /lib/firmware/intel/sof-ipc4-tplg/

UCM needs to be cloned and copied over to your target.

**Comment by falkanson** [comment]
Thanks for answering!
How do install a topology file? just copy the zib to the path?

**Comment by falkanson** [comment]
I now followed yours and brentlus instructions, the only differecne is that i installed the kernel from https://kernel.ubuntu.com/mainline/v6.10.6/

I still only have dummy audio, this is the dmesg output:

```
[    2.684887] i915 0000:00:02.0: [drm] Protected Xe Path (PXP) protected content support initialized
[    2.695999] [drm] Initialized i915 1.6.0 20230929 for 0000:00:02.0 on minor 1
[    2.697995] ACPI: video: Video Device [GFX0] (multi-head: yes  rom: no  post: no)
[    2.698193] iwlwifi 0000:00:14.3: Detected RF GF, rfid=0x2010d000
[    2.698406] input: Video Bus as /devices/LNXSYSTM:00/LNXSYBUS:00/PNP0A08:00/LNXVIDEO:00/input/input13
[    2.698722] sof-audio-pci-intel-mtl 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if 0x040100
[    2.698799] sof-audio-pci-intel-mtl 0000:00:1f.3: bound 0000:00:02.0 (ops i915_audio_component_bind_ops [i915])
[    2.705382] sof-audio-pci-intel-mtl 0000:00:1f.3: use msi interrupt mode
[    2.706510] fbcon: i915drmfb (fb0) is primary device
[    2.706513] fbcon: Deferring console take-over
[    2.706516] i915 0000:00:02.0: [drm] fb0: i915drmfb frame buffer device
[    2.721732] sof-audio-pci-intel-mtl 0000:00:1f.3: DMICs detected in NHLT tables: 2
[    2.721739] sof-audio-pci-intel-mtl 0000:00:1f.3: hda codecs found, mask 4
[    2.726793] sof-audio-pci-intel-mtl 0000:00:1f.3: SOF firmware and/or topology file not found.
[    2.726799] sof-audio-pci-intel-mtl 0000:00:1f.3: Supported default profiles
[    2.726801] sof-audio-pci-intel-mtl 0000:00:1f.3: - ipc type 1 (Requested):
[    2.726803] sof-audio-pci-intel-mtl 0000:00:1f.3:  Firmware file: intel/sof-ipc4/mtl/sof-mtl.ri
[    2.726804] sof-audio-pci-intel-mtl 0000:00:1f.3:  Topology file: intel/sof-ace-tplg/sof-mtl-rt712-l0-2ch.tplg
[    2.726806] sof-audio-pci-intel-mtl 0000:00:1f.3: Check if you have 'sof-firmware' package installed.
[    2.726807] sof-audio-pci-intel-mtl 0000:00:1f.3: Optionally it can be manually downloaded from:
[    2.726808] sof-audio-pci-intel-mtl 0000:00:1f.3:    https://github.com/thesofproject/sof-bin/
```

I really appreciate your help...

**Comment by plbossart** [comment]
@bardliao can you copy the topology file again, the instructions above are difficult to follow.

**Comment by falkanson** [comment]
Small update i got the speakers, no mic and not working, to show up by putting the file from [here](https://github.com/thesofproject/linux/issues/4923#issuecomment-2264421663) into /lib/firmware/intel/sof-ace-tplg/

**Comment by brentlu** [comment]
You need to copy sof-mtl-rt712-l0-2ch.tplg to /lib/firmware/intel/sof-ace-tplg

```
[    2.726804] sof-audio-pci-intel-mtl 0000:00:1f.3:  Topology file: intel/sof-ace-tplg/sof-mtl-rt712-l0-2ch.tplg
```

And you could find FW here if not installed in your system.
https://github.com/thesofproject/sof-bin/tree/main/v2.9.x/sof-ipc4-v2.9/mtl

**Comment by bardliao** [comment]
> @bardliao can you copy the topology file again, the instructions above are difficult to follow.

[sof-mtl-rt712-l0-2ch-4ch.zip](https://github.com/user-attachments/files/16464721/sof-mtl-rt712-l0-2ch-4ch.zip)

**Comment by shumingfan** [comment]
@falkanson I upstreamed the UCM config already. (https://github.com/alsa-project/alsa-ucm-conf/pull/435)
Before merging configs, you should copy these configs into your system.

**Comment by falkanson** [comment]
I followed you guys instructions, dmsg seams to be ok:
```
[    2.745824] sof-audio-pci-intel-mtl 0000:00:1f.3: DSP detected with PCI class/subclass/prog-if 0x040100
[    2.745978] sof-audio-pci-intel-mtl 0000:00:1f.3: bound 0000:00:02.0 (ops i915_audio_component_bind_ops [i915])
[    2.752934] sof-audio-pci-intel-mtl 0000:00:1f.3: use msi interrupt mode
[    2.753526] fbcon: i915drmfb (fb0) is primary device
[    2.753530] fbcon: Deferring console take-over
[    2.753533] i915 0000:00:02.0: [drm] fb0: i915drmfb frame buffer device
[    2.758177] iwlwifi 0000:00:14.3: Detected RF GF, rfid=0x2010d000
[    2.766800] sof-audio-pci-intel-mtl 0000:00:1f.3: DMICs detected in NHLT tables: 2
[    2.766804] sof-audio-pci-intel-mtl 0000:00:1f.3: hda codecs found, mask 4
[    2.770184] sof-audio-pci-intel-mtl 0000:00:1f.3: Firmware paths/files for ipc type 1:
[    2.770187] sof-audio-pci-intel-mtl 0000:00:1f.3:  Firmware file:     intel/sof-ipc4/mtl/sof-mtl.ri
[    2.770189] sof-audio-pci-intel-mtl 0000:00:1f.3:  Firmware lib path: intel/sof-ipc4-lib/mtl
[    2.770190] sof-audio-pci-intel-mtl 0000:00:1f.3:  Topology file:     intel/sof-ace-tplg/sof-mtl-rt712-l0-2ch.tplg
[    2.771245] sof-audio-pci-intel-mtl 0000:00:1f.3: Loaded firmware library: ADSPFW, version: 2.9.0.1
[    2.824528] iwlwifi 0000:00:14.3: base HW address: 04:e8:b9:37:34:bc
[    2.840978] iwlwifi 0000:00:14.3 wlp0s20f3: renamed from wlan0
[    2.876189] i915 0000:00:02.0: [drm] GT1: Loaded GSC firmware i915/mtl_gsc_1.bin (cv1.0, r102.0.0.1655, svn 1)
[    2.894461] sof-audio-pci-intel-mtl 0000:00:1f.3: Booted firmware version: 2.9.0.1
[    2.896292] i915 0000:00:02.0: [drm] GT1: HuC: authenticated for all workloads
[    2.900157] sof-audio-pci-intel-mtl 0000:00:1f.3: Topology: ABI 3:29:0 Kernel ABI 3:23:0
[    2.900544] sof_sdw sof_sdw: ASoC: Parent card not yet available, widget card binding deferred
[    2.915233] sof_sdw sof_sdw: hda_dsp_hdmi_build_controls: no PCM in topology for HDMI converter 3
```

but again no audio, while speakers show up but mic doesn't...
Speakers and Mic don't work - port works

Could the difference be that i got the 6.10.6 Kernel from another source?
What do i have to do with  `sof-mtl-rt712-l0-4ch.tplg`?

Additionally i get this output from `alsaucm reload`
```
ALSA lib parser.c:270:(parse_syntax_field) Incompatible syntax 7 in sof-soundwire.conf
ALSA lib main.c:1554:(snd_use_case_mgr_open) error: failed to import hw:0 use case configuration -22
alsaucm: error failed to open sound card hw:0: Invalid argument
```
Thanks again for helping me out with this

**Comment by marplusz** [comment]
@falkanson everything is described here
https://github.com/thesofproject/linux/issues/4923#issuecomment-2270519106
Have you installed alsa-ucm-conf and alsa-lib from the appropriate branches?

It works :)

**Comment by falkanson** [comment]
Im 99% sure i did, i just redid the entire process from a fresh Ubuntu 24.04.1 LTS Install and came to the same result

With these files: [AcerAudioFiles.zip](https://github.com/user-attachments/files/16785100/AcerAudioFiles.zip) I correctly downloaded from the sources

And the kernel from here: https://kernel.ubuntu.com/mainline/v6.10.6/

I ran these commands during installation:

Install the Kernel: (installed the 4 amd4 files)
```
sudo dpkg -i *.deb
```

Set the cum2 config
```
sudo rm -r /usr/share/alsa/ucm2
sudo cp -r ucm2 /usr/share/alsa
```

install alsa lib and dependencies for installation
```
sudo apt install libtool
sudo apt install make
./gitcompile
sudo make install
```

Copy the Topography and the firmware
```
sudo cp sof-mtl-rt712-l0-2ch.tplg /lib/firmware/intel/sof-ace-tplg/

sudo rm -r /lib/firmware/intel/sof-ipc4/mtl
sudo cp -r mtl /lib/firmware/intel/sof-ipc4/mtl/
```

reboot

I must have made a mistake somewhere if it is supposed to work... but i can't figure out where.

**Comment by marplusz** [comment]
@falkanson My installation process looks like this:

```
sudo cp /sof-mtl-rt712-l0-2ch-4ch/sof-mtl-rt712-l0-* /usr/lib/firmware/intel/sof-ipc4-tplg/
git clone https://github.com/shumingfan/alsa-ucm-conf.git
cd alsa-ucm-conf
git checkout add-rt712va
sudo cp -r ucm2/* /usr/share/alsa/ucm2/
sudo cp -r ucm/* /usr/share/alsa/ucm/                 //Now I saw that there is only a readme file :)
git clone https://github.com/alsa-project/alsa-lib.git
cd alsa-lib/
./gitcompile 
```
Here I see that you are copying topologies to another place:
I: sudo cp /sof-mtl-rt712-l0-2ch-4ch/sof-mtl-rt712-l0-* /usr/lib/firmware/intel/sof-ipc4-tplg/
You: sudo cp sof-mtl-rt712-l0-2ch.tplg /lib/firmware/intel/sof-ace-tplg/

Copy the topology files to the sof-ipc4-tplg directory. 
I didn't copy anything to /lib/firmware/intel/sof-ipc4-tplg/ and I see that these files are there too

**Comment by falkanson** [comment]
Weird, i followed you exact instructions and i'm back to only dummy audio, i at least have the jack working and the card showing up when copying to sof-ace-tplg
One thing to add is that i needed to create sof-ipc4-tplg
Sorry for being so difficult with this...

**Comment by marplusz** [comment]
@falkanson Did you copy both topology files? 
```
sudo find / | grep sof-mtl-rt712-l0-

/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-rt1712-l3.tplg.xz
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-2ch.tplg
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-4ch.tplg
```

**Comment by falkanson** [comment]
yes i have, both are in sof-ace-tplg and in sof-ipc4-tplg. Also i also tested it with them only being in one of the two locations. The speakers and mic never worked, jack only when they where in sof-ace-tplg.

```
find / | grep sof-mtl-rt712-
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-4ch.tplg
/usr/lib/firmware/intel/sof-ipc4-tplg/sof-mtl-rt712-l0-2ch.tplg
/usr/lib/firmware/intel/sof-ace-tplg/sof-mtl-rt712-l0-4ch.tplg
/usr/lib/firmware/intel/sof-ace-tplg/sof-mtl-rt712-l0-rt1712-l3.tplg.zst
/usr/lib/firmware/intel/sof-ace-tplg/sof-mtl-rt712-l0-2ch.tplg
/home/falkan/Downloads/ForAudio/sof-mtl-rt712-l0-2ch-4ch
/home/falkan/Downloads/ForAudio/sof-mtl-rt712-l0-2ch-4ch/sof-mtl-rt712-l0-4ch.tplg
/home/falkan/Downloads/ForAudio/sof-mtl-rt712-l0-2ch-4ch/sof-mtl-rt712-l0-2ch.tplg
```

**Comment by marplusz** [comment]
@falkanson  the only thing I can suggest is to install Fedora 40. It currently has Kernel version 6.10 and after adding the solutions presented here it should work.
There's probably a way to solve this for you, but I don't know Ubuntu.

**Comment by brentlu** [comment]
Maybe you could check if alsa-lib is updated successfully.

intel123@intel123-Swift-SFG14-72:~$ alsatplg -V
alsatplg version 1.2.9
libasound version 1.2.12
libatopology version 1.2.12

**Comment by falkanson** [comment]
it now wokrs after installing Fedora 40!
Thanks for helping me guys!

**Comment by marplusz** [comment]
Since kernel version 6.10.7 there is no sound on Fedora. I repeated the test in version 6.10.6 and there was sound. That's probably why @falkanson had problems on Ubuntu because  installed a newer kernel

**Comment by plbossart** [comment]
Known regression, see https://github.com/thesofproject/linux/issues/5168

This should be fixed shortly by linux-stable

if you can't wait, revert the commit "soundwire: stream: fix programming slave ports for non-continous port maps"

**Comment by VincentMTAC** [comment]
Quick comment, I have an Acer Swift Go 14, using Fedora 41 (kernel 6.11.3-300.fc41). 

Almost works out of the box, minus I had to download alsa-ucm-conf from [this repo](https://github.com/shumingfan/alsa-ucm-conf) (branch: add-rt712va) ucm2 folder and put it in my /usr/share/alsa/ucm2 one. 

Otherwise I had the speakers detected, no mic, and no sound. 

Thanks !

**Comment by SirSteel** [comment]
I have been bothered by this issue for a few weeks now. Tried so many things I lost track.
Today I installed Ubuntu 24.10 with the latest kernel 6.11.0-0-8.  (No improvement)
Then I installed latest alsa following these type of guides: https://www.linuxfromscratch.org/blfs/view/svn/multimedia/alsa-utils.html (No improvement)
And now I tried what Vincent suggested since I have the same computer. Used the repo and the branch.

Aaaand finally! It worked, I have sound, I have a mic. Hell yea!

Thanks mate!

**Comment by shassard** [comment]
On an updated Arch Linux machine, we have almost all the bits that we need for this to work with the exception of the patch to `alsa-ucm-conf` that hasn't been merged upstream.

I was able to get this to work with a one-liner to my system:
```shell
cd /usr/share/alsa && curl -q "https://github.com/alsa-project/alsa-ucm-conf/commit/9e3312c1471b8cb64036a5bd740e5c4f8c8f1186.diff" | sudo patch -p1
```

Then reboot the system.

**Comment by shassard** [comment]
I've created a PR for nixpkgs, which will hopefully get merged soon:
https://github.com/NixOS/nixpkgs/pull/355732

**Comment by SirSteel** [comment]
There was recently an update, which again destroyed my sound. I knew I should not have updated it but tried it anyways but now I am stuck again trying to get my sound to work.

But the main repo seems to have updated, so I just checked out master https://github.com/alsa-project/alsa-ucm-conf
and copied ucm and ucm2. And followed their restart instructions.

**Comment by liquidev** [comment]
I'm having the same problem; updated my packages, and sound is not working anymore.

What can I do to help diagnose the issue?

**Comment by bardliao** [comment]
@liquidev Please add this file
[sof-dyndbg.conf.txt](https://github.com/thesofproject/linux/files/14962935/sof-dyndbg.conf.txt) as /etc/modprobe.d/sof-dyndbg.conf, reboot then attach the dmesg log.

**Comment by liquidev** [comment]
Here's the verbose log after a fresh boot: [dmesg.txt](https://github.com/user-attachments/files/18255195/dmesg.txt)

**Comment by bardliao** [comment]
@liquidev 
```
Dec 26 20:59:32 construct kernel: soundwire sdw-master-0-0: sdw_compute_bus_params: could not find frame configuration for bus dr_freq 9600000
Dec 26 20:59:32 construct kernel: soundwire sdw-master-0-0: Compute params failed: -22
Dec 26 20:59:32 construct kernel:  SDW0-Capture-SimpleJack: ASoC: error at snd_soc_link_prepare on SDW0-Capture-SimpleJack: -22
```
It looks like the soundwire bus bandwidth is not sufficient in your configuration. I suspect that your topology uses 32 bit format.
Can you get [sof-mtl-rt712-l0-2ch.tplg](https://github.com/thesofproject/sof-bin/blob/main/v2.11.x/sof-ipc4-tplg-v2.11.3/sof-mtl-rt712-l0-2ch.tplg) from https://github.com/thesofproject/sof-bin/tree/main/v2.11.x/sof-ipc4-tplg-v2.11.3 and put it in the /lib/firmware/intel/sof-ace-tplg/sof-mtl-rt712-l0-2ch.tplg folder?
I.e Replace the sof-mtl-rt712-l0-2ch.tplg file in your device.

**Comment by liquidev** [comment]
@bardliao That helped, I have my headphone sound back, and now speakers are working too :3

Edit: just checked and mic is working too. Not sure about HDMI as I'm outside home right now and can't check.

**Comment by ujfalusi** [comment]
@rodrigovivi (@liquidev ), based on the comments this is now considered to be fixed? If so, can you close the issue?

**Comment by rodrigovivi** [comment]
I confirm that with recent kernel: 6.14.0-rc4+ and [sof-mtl-rt712-l0-2ch.tplg](https://github.com/thesofproject/sof-bin/blob/main/v2.11.x/sof-ipc4-tplg-v2.11.3/sof-mtl-rt712-l0-2ch.tplg) from https://github.com/thesofproject/sof-bin/tree/main/v2.11.x/sof-ipc4-tplg-v2.11.3, both built-in speaker and built-in microphone works. 

Thank you so much
